File: app/src/main/java/com/shevelev/comics_viewer/activities/main/comics_filters/RecentComicsFilter.java
package com.shevelev.comics_viewer.activities.main.comics_filters;


import com.shevelev.comics_viewer.activities.main.ComicsSortInfo;
import com.shevelev.comics_viewer.common.helpers.CollectionsHelper;
import com.shevelev.comics_viewer.dal.DalFacade;
import com.shevelev.comics_viewer.dal.dto.Comics;

import java.util.List;

/**
 * Get recent comics - public and private
 */
public class RecentComicsFilter extends FilterBase
{
    private static final int comicsToReturn=6;

    public RecentComicsFilter(ComicsSortInfo comicsSortInfo, boolean isPrivateComicsHidden)
    {
        super(comicsSortInfo, isPrivateComicsHidden);
    }

    @Override
    protected List<Comics> getComicsList()
    {
        List<Comics> result =  DalFacade.Comics.getComics(true);

        result = CollectionsHelper.where(result, item->item.lastViewDate!=null);
        result = CollectionsHelper.sort(result, (c1, c2) -> c1.lastViewDate.compareTo(c2.lastViewDate), true);
        result = CollectionsHelper.take(result, comicsToReturn, 0);

        return result;
    }
}

File: app/src/main/java/com/shevelev/comics_viewer/activities/main/comics_filters/RecentComicsFilter.kt
package com.shevelev.comics_viewer.activities.main.comics_filters

import android.content.Context
import com.shevelev.comics_viewer.activities.main.ComicsSortInfo
import com.shevelev.comics_viewer.utils.filterOrNull
import com.shevelev.comics_viewer.utils.sort
import com.shevelev.comics_viewer.utils.takeOrNull
import com.shevelev.comics_viewer.dal.DalFacade
import com.shevelev.comics_viewer.dal.dto.Comics

/**
 * Get recent comics - public and private
 */
class RecentComicsFilter(
    context: Context,
    comicsSortInfo: ComicsSortInfo?,
    isPrivateComicsHidden: Boolean
) : FilterBase(context, comicsSortInfo, isPrivateComicsHidden) {
    override val comicsList: List<Comics>
        protected get() {
            var result = DalFacade.Comics.getComics(true)

            result = result.filterOrNull { item: Comics -> item.lastViewDate != null }!!

            val comparator = Comparator { c1: Comics, c2: Comics -> c1.lastViewDate?.compareTo(c2.lastViewDate) ?: 0 }

            result = result.sort(comparator, true)
            result = result.takeOrNull(comicsToReturn, 0)
            return result!!
        }

    companion object {
        private const val comicsToReturn = 6
    }
}